### ISSUE TITLE
[W10] Settings doesn't stick on restart

### ISSUE THREAD

**Issue by Porkepix** [mumble-voip/mumble#6628]
### Description

On Windows (at least on 10), trying to change the language to a different from default one doesn't work (here it was specifically to switch it to English to have more accurate reports of two other bugs for Mumble).

See the STR below, but it seems impossible to change the language: setting doesn't stick after restart.

### Steps to reproduce

1. Go to the language selector in the configuration and select another one than default (I tried with both American and British English, I doubt they would be the only affected ones).
2. Clicking 'apply' does nothing
3. Clicking 'Ok' close the configuration window, and opens a popup asking to restart Mumble, as changing language requires that
4. Accept restart
5. Language didn't change, and if you check in the configuration, it's back at default
6. I tries with only apply and closing Mumble by hand rather than restarting from the popup, result is the same.

### Mumble version

1.5.634

### Mumble component

Client

### OS

Windows

### Reproducible?

Yes

### Relevant log output

```shell
I have no idea on how to gather logs from Windows, as opposed to just open this from terminal and/or checking journald for Linux.
```

### Screenshots

N/A

**Comment by davidebeatrici** [comment]
Are other settings applied correctly?

**Comment by davidebeatrici** [comment]
@Krzmbrzl This could be potentially related to #6567.

**Comment by Hartmnt** [comment]
@davidebeatrici Agreed, but this only seems to affect a handful of users. People in our Matrix room are reporting the settings are being saved just fine. The question now becomes: What is the common condition that connects all users with issues.

**Comment by Hartmnt** [comment]
@Porkepix Would you be able to provide us with logs from the [Developer Console](https://github.com/mumble-voip/mumble/blob/master/docs/DeveloperConsole.md)?

**Comment by Hartmnt** [comment]
@Porkepix Could you please check, if the folders in this path exist? Specifically the Mumble related folders. ``C:/Users/$USER/AppData/Local/Mumble/Mumble/``

The public Matrix room is here: https://matrix.to/#/#mumble:matrix.org

**Comment by Krzmbrzl** [comment]
How does a path with `$USER` in it exist? This seems like something the shell would have to expand but I doubt that the way we access the path, any shell expansion is performed. To me it seems like what we really should have is the actual user name instead of `$USER` 👀

**Comment by Krzmbrzl** [comment]
Ah okay. Now that makes sense :+1:

**Comment by pateus** [comment]
Similar problem, settings are not saved, have to configure again each time you log in. Running as administrator does not help.
`<W>2024-11-24 00:40:29.283 Failed at moving settings from C:/Users/Пользователь/AppData/Local/Temp/mumble_settings.json.tmp to C:/Users/Пользователь/AppData/Local/Mumble/Mumble/mumble_settings.json - reason: Исходный файл отсутствует.`
There is no mumble_settings.json.tmp file in directory C:/Users/Пользователь/AppData/Local/Temp/, directory  C:/Users/Пользователь/AppData/Local/Mumble/Mumble/ is there but empty.

**Comment by Hartmnt** [comment]
Ok, wild punt, but does everyone affected by this issue have a "special character" in their path? As in, does your username contain one or more non-ASCII characters? Ping @Porkepix @Mihail-Lazarev @markusmueller73

**Comment by markusmueller73** [comment]
@Hartmnt Same here, have an umlaut in my username.

**Comment by Hartmnt** [comment]
Nice, that gives us a reasonable assumption of what goes wrong here and why this issue is relatively rare.

We let you know when we have a bugfix that you guys can test before we merge it, since we (the Mumble core team) can't reproduce this easily.

**Comment by pateus** [comment]
The problem most likely occurs with the username in the native language. For example, the ASCII table does not have Russian (Cyrillic) letters.

**Comment by Hartmnt** [comment]
Hi all, I need someone who can test this experimental build: https://dev.azure.com/Mumble-VoIP/c819eb06-7b22-4ef3-bbcd-860094454eb3/_apis/build/builds/8990/artifacts?artifactName=Windows%20%28x64%29%20installer&api-version=7.1&%24format=zip

Please install the version of Mumble linked above and let me know, if your settings now stick. If they do not stick, please provide the output of the [Developer Console](https://github.com/mumble-voip/mumble/blob/master/docs/DeveloperConsole.md).

Also note that this is from the unstable 1.6.x branch, so I would not recommend this as a daily driver.

If you let me know ASAP, I might be able to get this into the next 1.5.x release.

Ping @Porkepix @pateus @mfxbe @markusmueller73

**Comment by Hartmnt** [comment]
> If we upgrade to this 1.6.x, is it able to go back to 1.5.x after that without issues, or we need to somehow backup some files before and restore them after?

You will probably have to fully uninstall 1.6.x as the Windows Installer does not allow software to be downgraded IIRC.

So, definitely backup any data related to Mumble that you care for. The downgrade could break things and even the upgrade may possibly break something as 1.6.x is not that thoroughly tested yet.

> Also, one things I mentioned but that wasn't really visible: the settings not saving were a regression (but from when, I don't know), considering I had one shortcut saved in the past, and that one was still sticking around, but no way to add other ones.

The Mumble devs current collective assumption is that a recent Windows Update somehow broke how we were handling paths. But I can not say for sure. We all use Linux as a daily driver.

> Saving servers is still working, too.

There we are interacting with a sqlite database with the proper abstraction in place. The settings file is a raw file output stream we write to.

**Comment by Hartmnt** [comment]
> mentioning that my (not the exact text, as it's in French) "config file references a database (sqlite) file but that this file doesn't exist.

What does it say in french? It is easier to look up where that string comes from in the original language.

> Where this config file could be? What's its name? I was not able to find one located at another place than those two directories.

Are you providing a config file via ``--config`` to the application launcher?

Is there a ``mumble_settings.json`` in your backup?

If you are not using ``--config`` or ``-c`` and ``mumble_settings.json`` exists in your backup, I think you are seeing this message by mistake.

> Also, another weirdness I noticed while looking for it: Mumble install is located into `C:\Programs\Mumble`. In there I found a `Versions` directory containing another directory that's probably the version I installed in 2021: `1.3.4`. Is there any reason to locate a big part of Mumble's files in a versioned directory sticking to the number of the first installed version? It was the only one, no `1.5.x` directory in there.

No idea. I assume the way we install on Windows has changed in the last 5 years, so I guess you are seeing part of the old process.

**Comment by pateus** [comment]
The error remains.
`<W>2024-12-02 21:33:50.982 Failed at writing temporary settings file: C:/Users/Пользователь/AppData/Local/Temp/mumble_settings.json.tmp
<W>2024-12-02 21:33:50.982 Failed at moving settings from C:/Users/Пользователь/AppData/Local/Temp/mumble_settings.json.tmp to C:/Users/Пользователь/AppData/Local/Mumble/Mumble/mumble_settings.json - reason: Исходный файл отсутствует.`

**Comment by Hartmnt** [comment]
> The error remains. `<W>2024-12-02 21:33:50.982 Failed at writing temporary settings file: C:/Users/Пользователь/AppData/Local/Temp/mumble_settings.json.tmp <W>2024-12-02 21:33:50.982 Failed at moving settings from C:/Users/Пользователь/AppData/Local/Temp/mumble_settings.json.tmp to C:/Users/Пользователь/AppData/Local/Mumble/Mumble/mumble_settings.json - reason: Исходный файл отсутствует.`

Damn, but thank you very much for testing.

**Comment by pateus** [comment]
> > The error remains. `<W>2024-12-02 21:33:50.982 Failed at writing temporary settings file: C:/Users/Пользователь/AppData/Local/Temp/mumble_settings.json.tmp <W>2024-12-02 21:33:50.982 Failed at moving settings from C:/Users/Пользователь/AppData/Local/Temp/mumble_settings.json.tmp to C:/Users/Пользователь/AppData/Local/Mumble/Mumble/mumble_settings.json - reason: Исходный файл отсутствует.`
> 
> Damn, but thank you very much for testing.

Maybe it makes sense to make it possible to choose the path for installing the application?

**Comment by pateus** [comment]
I deleted the old version and installed the new one, there was no database connection error, the server for connection was pulled up from the old version.

**Comment by Hartmnt** [comment]
> Maybe it makes sense to make it possible to choose the path for installing the application?

This is not an option sadly, we need to work around how Windows handles encoding internally.

I will be back shortly with another attempt.

> No, but I think it's kind of the point of this issue: it looks like this file never was created for me when you moved the configs to a `.json` file.

Ah yes, totally makes sense. When you upgraded from 1.3 to 1.4 or somthing the old registry keys were not deleted. Now that the json/sqlite files are gone, it probably defaults to registry keys or something like that. I think you can ignore this and just create a new database for testing.

**Comment by Hartmnt** [comment]
> > 
> 
> Am I to consider the registry keys were used pre-1.4 and are now all obsoletes, or are some still used? (Bonus question, if the answer is positive: is Mumble's setup removing those if we use a clean uninstall with it?

Yes, I guess it would  have been smart to remove the registry keys on a successful migration to json



> It also seems that now, if I try to change the language, Mumble ask for a restart as usual, and when I answer yes, it keeps a zombie process around (without systray icon), preventing to launch Mumble anymore until I go to kill the process manually.

Yes, this is already tracked by #4932 and  will be fixed with the PR that rewrites the tray icon implementation

**Comment by Hartmnt** [comment]
> Can those registry keys be the culprit for the last popup I posted, about shortcuts? Or it's completely unrelated? My Mumble seems to be in a quite weird state right now.

Most likely related. You can disregard this as Mumble is trying to migrate your registry keys again into a json file. When you move your backup files back, those take priority.

**Comment by Hartmnt** [comment]
Alright, I have a new build to test

https://dev.azure.com/Mumble-VoIP/c819eb06-7b22-4ef3-bbcd-860094454eb3/_apis/build/builds/8991/artifacts?artifactName=Windows%20%28x64%29%20installer&api-version=7.1&%24format=zip

Ping @Porkepix @pateus

**Comment by mfxbe** [comment]
> Hi all, I need someone who can test this experimental build: https://dev.azure.com/Mumble-VoIP/c819eb06-7b22-4ef3-bbcd-860094454eb3/_apis/build/builds/8990/artifacts?artifactName=Windows%20%28x64%29%20installer&api-version=7.1&%24format=zip
> 
> Please install the version of Mumble linked above and let me know, if your settings now stick. If they do not stick, please provide the output of the [Developer Console](https://github.com/mumble-voip/mumble/blob/master/docs/DeveloperConsole.md).
> 
> Also note that this is from the unstable 1.6.x branch, so I would not recommend this as a daily driver.
> 
> If you let me know ASAP, I might be able to get this into the next 1.5.x release.
> 
> Ping @Porkepix @pateus @mfxbe @markusmueller73

Unfortunately I can't test the builds currently. I only have Linux systems as well and the affected Windows laptop left town with my SO.

Will do it as soon as possible (if testing is still needed then.)

**Comment by Hartmnt** [comment]
> * I got rid of both and installed your test build (a question here, what's the PDB's folder in here? I was surprised by the size of the archive)

Debug symbols iirc

> * is still around and appeared, I don't understand why

The registry keys are probably not deleted on uninstall

> * Finally I couldn't end the test because I have the main window and the consent window that starts to draw and I get the outline, then it goes in a "does not respond" state and get crashed/killed by the OS a couple of seconds later.

Possibly still related to the tray icon thing. Could you try again and if this crash keeps happening, start Mumble in a CMD and see if that prints anything useful, please?

**Comment by pateus** [comment]
> Хорошо, мне нужно протестировать новую сборку.
> 
> [https://dev.azure.com/Mumble-VoIP/c819eb06-7b22-4ef3-bbcd-860094454eb3/_apis/build/builds/8991/artifacts?artifactName=Windows%20%28x64%29%20installer&api-version=7.1& %24format=zip](https://dev.azure.com/Mumble-VoIP/c819eb06-7b22-4ef3-bbcd-860094454eb3/_apis/build/builds/8991/artifacts?artifactName=Windows%20%28x64%29%20installer&api-version=7.1&%24format=zip)
> 
> Пинг@Porkepix @pateus

Thank you. Now everything works correctly. The settings are saved.

**Comment by Hartmnt** [comment]
> Thank you. Now everything works correctly. The settings are saved.

Very cool. Thanks for testing, so that this can land in the next release :+1: 

@Porkepix At this point, I would recommend uninstalling Mumble completely again and manually removing the settings/sqlite files located in ``%AppData%``

**Comment by Hartmnt** [comment]
> I wanted to verify the issue is solved with the newly released `1.5.735`, unfortunately, that release completely crash on startup the same way the release from [#6628 (comment)](https://github.com/mumble-voip/mumble/issues/6628#issuecomment-2512614919) did and I mentioned in [#6628 (comment)](https://github.com/mumble-voip/mumble/issues/6628#issuecomment-2513252699)
> 
> Any issue related? Any way to get debug information?

Ok, have you tried upgrading from 1.5.634 or did you delete everything before installing 1.5.735?
